File: backend/app/schemas/admin_content.py
from datetime import date as date_
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.public import GuestIn

SLUG_RE = r"^[a-z0-9]+(?:-[a-z0-9]+)*$"


class PropertyCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    slug: str = Field(min_length=1, max_length=150, pattern=SLUG_RE)
    city: str = Field(min_length=1, max_length=100)
    state: str = Field(min_length=1, max_length=100)
    description: str = Field(min_length=1, max_length=4000)
    cover_image_url: str = Field(min_length=1, max_length=2000)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    amenities: list[str] = Field(default_factory=list, max_length=30)

    @field_validator("amenities")
    @classmethod
    def clean_amenities(cls, v: list[str]) -> list[str]:
        cleaned = [a.strip()[:60] for a in v if a.strip()]
        if len(cleaned) != len(v):
            raise ValueError("Amenity labels cannot be blank")
        return cleaned


class PropertyUpdate(BaseModel):
    """All fields optional — only what's supplied gets patched (never id/timestamps)."""

    name: str | None = Field(default=None, min_length=1, max_length=150)
    slug: str | None = Field(default=None, min_length=1, max_length=150, pattern=SLUG_RE)
    city: str | None = Field(default=None, min_length=1, max_length=100)
    state: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, min_length=1, max_length=4000)
    cover_image_url: str | None = Field(default=None, min_length=1, max_length=2000)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    amenities: list[str] | None = Field(default=None, max_length=30)


class RoomTypeCreate(BaseModel):
    property_id: int = Field(gt=0)
    name: str = Field(min_length=1, max_length=150)
    description: str = Field(min_length=1, max_length=4000)
    max_occupancy: int = Field(gt=0, le=20)
    base_price: Decimal = Field(gt=0, le=1_000_000)
    images: list[str] = Field(default_factory=list, max_length=30)


class RoomTypeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    description: str | None = Field(default=None, min_length=1, max_length=4000)
    max_occupancy: int | None = Field(default=None, gt=0, le=20)
    base_price: Decimal | None = Field(default=None, gt=0, le=1_000_000)
    images: list[str] | None = Field(default=None, max_length=30)


class RatePlanCreate(BaseModel):
    room_type_id: int = Field(gt=0)
    name: str = Field(min_length=1, max_length=100)
    fixed_price: Decimal = Field(gt=0, le=1_000_000)
    refundable: bool = True
    includes_breakfast: bool = False
    valid_from: date_
    valid_to: date_

    @field_validator("valid_to")
    @classmethod
    def validate_range(cls, v: date_, info) -> date_:
        valid_from = info.data.get("valid_from")
        if valid_from and v <= valid_from:
            raise ValueError("valid_to must be after valid_from")
        return v


class RatePlanUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    fixed_price: Decimal | None = Field(default=None, gt=0, le=1_000_000)
    refundable: bool | None = None
    includes_breakfast: bool | None = None
    valid_from: date_ | None = None
    valid_to: date_ | None = None


class AvailabilityDayIn(BaseModel):
    date: date_
    rooms_available: int = Field(ge=0, le=50)


class AvailabilityBulkUpsert(BaseModel):
    days: list[AvailabilityDayIn] = Field(min_length=1, max_length=120)


class GuestVerifyIn(BaseModel):
    status: Literal["verified", "rejected"]
    note: str | None = Field(default=None, max_length=500)


class AdminBookingCreate(BaseModel):
    """A staff-entered booking for a walk-in/phone reservation. property_id is
    accepted here but a scoped property_manager's own property always wins —
    see scope_property_id in routers/admin/bookings.py."""

    property_id: int = Field(gt=0)
    room_type_id: int = Field(gt=0)
    rate_plan_id: int = Field(gt=0)
    check_in: date_
    check_out: date_
    guests_count: int = Field(gt=0, le=20)
    guest: GuestIn

    @field_validator("check_out")
    @classmethod
    def validate_range(cls, v: date_, info) -> date_:
        check_in = info.data.get("check_in")
        if check_in and v <= check_in:
            raise ValueError("check_out must be after check_in")
        return v


class PackageCreate(BaseModel):
    property_id: int = Field(gt=0)
    name: str = Field(min_length=1, max_length=150)
    nights: int = Field(gt=0, le=30)
    description: str = Field(min_length=1, max_length=2000)
    price: Decimal = Field(gt=0, le=1_000_000)
    meal_plan: str = Field(min_length=1, max_length=100)
    rooms_included: int = Field(default=1, gt=0, le=20)
    max_guests: int = Field(gt=0, le=40)
    cover_image_url: str | None = Field(default=None, max_length=2000)
    active: bool = True


class PackageUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=150)
    nights: int | None = Field(default=None, gt=0, le=30)
    description: str | None = Field(default=None, min_length=1, max_length=2000)
    price: Decimal | None = Field(default=None, gt=0, le=1_000_000)
    meal_plan: str | None = Field(default=None, min_length=1, max_length=100)
    rooms_included: int | None = Field(default=None, gt=0, le=20)
    max_guests: int | None = Field(default=None, gt=0, le=40)
    cover_image_url: str | None = Field(default=None, max_length=2000)
    active: bool | None = None


class HeroContentIn(BaseModel):
    """Loosely-typed content-management payload — capped to keep it from
    being used to stash arbitrarily large blobs in the DB."""

    headline: str | None = Field(default=None, max_length=200)
    subtext: str | None = Field(default=None, max_length=1000)
    cta_label: str | None = Field(default=None, max_length=60)


class AboutValueIn(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    body: str = Field(min_length=1, max_length=500)


class AboutContentIn(BaseModel):
    headline: str | None = Field(default=None, max_length=200)
    intro: str | None = Field(default=None, max_length=1500)
    story: str | None = Field(default=None, max_length=1500)
    values: list[AboutValueIn] | None = Field(default=None, max_length=6)

File: backend/src/schemas/adminContent.ts
import { z } from "zod";
import { GuestInSchema } from "./public";

export const SLUG_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

export const PropertyCreateSchema = z.object({
  name: z.string().min(1).max(150),
  slug: z.string().min(1).max(150).regex(SLUG_RE),
  city: z.string().min(1).max(100),
  state: z.string().min(1).max(100),
  description: z.string().min(1).max(4000),
  cover_image_url: z.string().min(1).max(2000),
  latitude: z.number().min(-90).max(90).nullable().optional().default(null),
  longitude: z.number().min(-180).max(180).nullable().optional().default(null),
  amenities: z.array(z.string()).max(30).default([]),
});
export type PropertyCreate = z.infer<typeof PropertyCreateSchema>;

export const PropertyUpdateSchema = z.object({
  name: z.string().min(1).max(150).optional(),
  slug: z.string().min(1).max(150).regex(SLUG_RE).optional(),
  city: z.string().min(1).max(100).optional(),
  state: z.string().min(1).max(100).optional(),
  description: z.string().min(1).max(4000).optional(),
  cover_image_url: z.string().min(1).max(2000).optional(),
  latitude: z.number().min(-90).max(90).nullable().optional(),
  longitude: z.number().min(-180).max(180).nullable().optional(),
  amenities: z.array(z.string()).max(30).optional(),
});

export const RoomTypeCreateSchema = z.object({
  property_id: z.number().int().positive(),
  name: z.string().min(1).max(150),
  description: z.string().min(1).max(4000),
  max_occupancy: z.number().int().positive().max(20),
  base_price: z.number().positive().max(1_000_000),
  images: z.array(z.string()).max(30).default([]),
});

export const RoomTypeUpdateSchema = z.object({
  name: z.string().min(1).max(150).optional(),
  description: z.string().min(1).max(4000).optional(),
  max_occupancy: z.number().int().positive().max(20).optional(),
  base_price: z.number().positive().max(1_000_000).optional(),
  images: z.array(z.string()).max(30).optional(),
});

export const RatePlanCreateSchema = z
  .object({
    room_type_id: z.number().int().positive(),
    name: z.string().min(1).max(100),
    fixed_price: z.number().positive().max(1_000_000),
    refundable: z.boolean().default(true),
    includes_breakfast: z.boolean().default(false),
    valid_from: z.string().regex(DATE_RE),
    valid_to: z.string().regex(DATE_RE),
  })
  .refine((v) => v.valid_to > v.valid_from, { message: "valid_to must be after valid_from", path: ["valid_to"] });

export const RatePlanUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  fixed_price: z.number().positive().max(1_000_000).optional(),
  refundable: z.boolean().optional(),
  includes_breakfast: z.boolean().optional(),
  valid_from: z.string().regex(DATE_RE).optional(),
  valid_to: z.string().regex(DATE_RE).optional(),
});

export const AvailabilityDayInSchema = z.object({
  date: z.string().regex(DATE_RE),
  rooms_available: z.number().int().min(0).max(50),
});

export const AvailabilityBulkUpsertSchema = z.object({
  days: z.array(AvailabilityDayInSchema).min(1).max(120),
});

export const GuestVerifyInSchema = z.object({
  status: z.enum(["verified", "rejected"]),
  note: z.string().max(500).nullable().optional().default(null),
});

export const AdminBookingCreateSchema = z
  .object({
    property_id: z.number().int().positive(),
    room_type_id: z.number().int().positive(),
    rate_plan_id: z.number().int().positive(),
    check_in: z.string().regex(DATE_RE),
    check_out: z.string().regex(DATE_RE),
    guests_count: z.number().int().positive().max(20),
    guest: GuestInSchema,
  })
  .refine((v) => v.check_out > v.check_in, { message: "check_out must be after check_in", path: ["check_out"] });
export type AdminBookingCreate = z.infer<typeof AdminBookingCreateSchema>;

export const PackageCreateSchema = z.object({
  property_id: z.number().int().positive(),
  name: z.string().min(1).max(150),
  nights: z.number().int().positive().max(30),
  description: z.string().min(1).max(2000),
  price: z.number().positive().max(1_000_000),
  meal_plan: z.string().min(1).max(100),
  rooms_included: z.number().int().positive().max(20).default(1),
  max_guests: z.number().int().positive().max(40),
  cover_image_url: z.string().max(2000).nullable().optional().default(null),
  active: z.boolean().default(true),
});

export const PackageUpdateSchema = z.object({
  name: z.string().min(1).max(150).optional(),
  nights: z.number().int().positive().max(30).optional(),
  description: z.string().min(1).max(2000).optional(),
  price: z.number().positive().max(1_000_000).optional(),
  meal_plan: z.string().min(1).max(100).optional(),
  rooms_included: z.number().int().positive().max(20).optional(),
  max_guests: z.number().int().positive().max(40).optional(),
  cover_image_url: z.string().max(2000).nullable().optional(),
  active: z.boolean().optional(),
});

export const HeroContentInSchema = z.object({
  headline: z.string().max(200).nullable().optional(),
  subtext: z.string().max(1000).nullable().optional(),
  cta_label: z.string().max(60).nullable().optional(),
});

const AboutValueInSchema = z.object({
  title: z.string().min(1).max(100),
  body: z.string().min(1).max(500),
});

export const AboutContentInSchema = z.object({
  headline: z.string().max(200).nullable().optional(),
  intro: z.string().max(1500).nullable().optional(),
  story: z.string().max(1500).nullable().optional(),
  values: z.array(AboutValueInSchema).max(6).nullable().optional(),
});
